### PR TITLE
`__getstate__` and `__setstate__` consolidation

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
The `Process` class will now take care of all `__getstate__` and `__setstate__`

This PR will then ensure that any forked process, on windows, and of course all platforms which support forking, will behave properly **without** having to implement their own get and set state functions, unless they need to extend the base functionality provided in out Process class.

### Previous Behavior
All subclasses had to implement `__getstate__` and `__setstate__`

### New Behavior
The `Process` class will now take care of all `__getstate__` and `__setstate__`

### Tests written?

No

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/engines/__init__.py [hunk 1]
```diff
@@ -81,33 +81,6 @@ def __init__(self, opts, fun, config, funcs, runners, proxy, **kwargs):
...
-            'opts': self.opts,
-            'fun': self.fun,
-            'config': self.config,
-            'funcs': self.funcs,
-            'runners': self.runners,
-            'proxy': self.proxy,
-            'log_queue': self.log_queue,
-            'log_queue_level': self.log_queue_level
-        }
-
```
Comment (waynew) [inline salt/engines/__init__.py]: Not sure that I understand - how do all of these things end out getting called? They're quite different than what's in the other files.

Comment (s0undt3ch) [reply salt/engines/__init__.py]: On windows, and windows alone, the `__new__` method will keep a copy of the `*args` and `**kwargs` used to instantiate the class.
Then, again, only on windows, the get and set state methods will use that information to re-instsntiate the class.

Comment (s0undt3ch) [reply salt/engines/__init__.py]: [Here](https://github.com/saltstack/salt/pull/55793/files#diff-0eb908cc193e4bf8dd9f96567a3abfebR741)

**Comment by s0undt3ch** [comment]
This PR will then ensure that any forked process, on windows, and of course all platforms which support forking, will behave properly **without** having to implement their own get and set state functions.

**Comment by s0undt3ch** [comment]
Unless they need to extend the base functionality provided in out Process class.

**Review by twangboy — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
I need to make sure any extra methods for post fork and pre shutdown are also carried over to the new process.

**Comment by s0undt3ch** [comment]
re-run full all

**Review by dwoz — COMMENTED** [review]
I'm not a fan of this. It seems reasonable to me to expect sub-classes of Process to implement their own `__getstate__` and `__setstate__` rather than trying to be smart about it in the constructor.

**Comment by s0undt3ch** [comment]
As we discussed, it gets us, at least, logging on those processes for free.

**Review by twangboy — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
Conflicts resolved

**Review by dwoz — DISMISSED** [review]

**Review by krionbsd — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
The pre-commit issue is being solved in https://github.com/saltstack/salt/pull/60254

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

File: salt/utils/platform.py [hunk 2]
```diff
@@ -209,3 +210,12 @@ def is_aarch64():
     Simple function to return if host is AArch64 or not
     """
     return platform.machine().startswith("aarch64")
+
+
+def spawning_platform():
```
Comment (waynew) [inline salt/utils/platform.py:215]: **question** is it better to gate this behind a function or do something like this at the module level?

```
SPAWNING_PLATFORM = multiprocessing.get_start_method(allow_none=False) == "spawn"
```

I don't know if it's called enough to make a difference :shrug:

Comment (s0undt3ch) [reply salt/utils/platform.py:215]: No, the method can be changed programmatically at run time. A global variable would only pick the initial config value.

Comment (s0undt3ch) [reply salt/utils/platform.py:215]: @waynew are we good to resolve this review comment?

**Review by twangboy — APPROVED** [review]

File: salt/utils/platform.py (hunk 2 above)
Comment (s0undt3ch) [reply salt/utils/platform.py:215]: Also, if you notice, this is probably the only function which doesnt have the cached decorator, that was the reason behind it.

**Review by dwoz — APPROVED** [review]

**Pull request merged**